File: Yijing.and/app/src/main/java/org/yijing/CTrigramValueSequencer.java
package org.yijing;

public class CTrigramValueSequencer extends CValueSequencer
{
	public CTrigramValueSequencer(int nValue) {
		super(3, 8, nValue);
		m_pvsInner = new CLineValueSequencer[3];
		m_pvsInner[2] = new CLineValueSequencer(0);
		m_pvsInner[1] = new CLineValueSequencer(0);
		m_pvsInner[0] = new CLineValueSequencer(0);
		Line(2).SetParent(this);
		Line(1).SetParent(this);
		Line(0).SetParent(this);
		m_nSequences = Sequences.nTrigramSequences;
		m_nRatios = Sequences.nTrigramRatios;
	}

	public CLineValueSequencer Line(int nIndex) {
		return (CLineValueSequencer)m_pvsInner[nIndex];
	}

	public CValueSequencer Inverse() {
		boolean[] bMoving = { false, false, false };
		SaveMoving(bMoving);
		Value((Line(2).Value() % 2 == 0 ? 0 : 1) +
			((Line(1).Value() % 2 == 0 ? 0 : 1) * 2) +
			((Line(0).Value() % 2 == 0 ? 0 : 1) * 4));
		UpdateInnerValues();
		UpdateOuterValues();
		RestoreMoving(bMoving, true);
		return this;
	}

	public CValueSequencer Opposite() {
		boolean[] bMoving = { false, false, false };
		SaveMoving(bMoving);
		super.Opposite();
		RestoreMoving(bMoving, false);
		return this;
	}

	public CValueSequencer Move() {
		Line(2).Move();
		Line(1).Move();
		Line(0).Move();
		return this;
	}

	public CValueSequencer Young() {
		Line(2).Young();
		Line(1).Young();
		Line(0).Young();
		return this;
	}

	protected void UpdateInnerValues() {
		switch (m_nValue)
		{
			case 0:
				Line(2).Value(m_nYinLine);
				Line(1).Value(m_nYinLine);
				Line(0).Value(m_nYinLine);
				break;
			case 1:
				Line(2).Value(m_nYinLine);
				Line(1).Value(m_nYinLine);
				Line(0).Value(m_nYangLine);
				break;
			case 2:
				Line(2).Value(m_nYinLine);
				Line(1).Value(m_nYangLine);
				Line(0).Value(m_nYinLine);
				break;
			case 3:
				Line(2).Value(m_nYinLine);
				Line(1).Value(m_nYangLine);
				Line(0).Value(m_nYangLine);
				break;
			case 4:
				Line(2).Value(m_nYangLine);
				Line(1).Value(m_nYinLine);
				Line(0).Value(m_nYinLine);
				break;
			case 5:
				Line(2).Value(m_nYangLine);
				Line(1).Value(m_nYinLine);
				Line(0).Value(m_nYangLine);
				break;
			case 6:
				Line(2).Value(m_nYangLine);
				Line(1).Value(m_nYangLine);
				Line(0).Value(m_nYinLine);
				break;
			case 7:
				Line(2).Value(m_nYangLine);
				Line(1).Value(m_nYangLine);
				Line(0).Value(m_nYangLine);
				break;
		}
		Line(2).UpdateInnerValues();
		Line(1).UpdateInnerValues();
		Line(0).UpdateInnerValues();
	}

	protected void UpdateOuterValues() {
		Value((Line(0).Value() % 2 == 0 ? 0 : 1) +
			((Line(1).Value() % 2 == 0 ? 0 : 1) * 2) +
			((Line(2).Value() % 2 == 0 ? 0 : 1) * 4));
		if(m_pvsParent != null)
			m_pvsParent.UpdateOuterValues();
	}

	private void SaveMoving(boolean[] bMoving) {
		for (int l = 0; l < 3; ++l)
			if ((Line(l).Value() == 0) || (Line(l).Value() == 3))
				bMoving[l] = true;
	}

	private void RestoreMoving(boolean[] bMoving, boolean bInverseLine) {
		for (int l = 0; l < 3; ++l)
			if (bMoving[l])
			{
				int l1 = l;
				if (bInverseLine)
					if (l1 == 0)
						l1 = 2;
					else
						if (l1 == 2)
							l1 = 0;
				Line(l1).Old();
			}
	}

	public static void SetCurrentSequence(int nSequence) {
		m_nCurrentSequence = nSequence;
	}

	public static void SetCurrentRatio(int nRatio) {
		m_nCurrentRatio = nRatio;
	}
	
	public static void SetCurrentLabel(int nLabel) {
		m_nCurrentLabel = nLabel;
	}
	
	protected String GetLabel() { 
		return Sequences.strTrigramLabels[GetCurrentLabel()][Value()];
	}
	
	protected boolean GetMoving() {
		return Line(0).IsMoving() || Line(1).IsMoving() || Line(2).IsMoving();
	}

	protected int GetCurrentSequence() { return m_nCurrentSequence; }
	protected int GetCurrentRatio() { return m_nCurrentRatio; }
	protected int GetCurrentLabel() { return m_nCurrentLabel; }

	protected static int m_nCurrentSequence = 0;
	protected static int m_nCurrentRatio = 0;
	protected static int m_nCurrentLabel = 0;

	protected int m_nYinLine = 2;
	protected int m_nYangLine = 1;
}

File: ValueSequencer.ts/CTrigramValueSequencer.ts
import { CValueSequencer } from "./CValueSequencer";
import { CLineValueSequencer } from "./CLineValueSequencer";
import { Sequences } from "./Sequences";

export class CTrigramValueSequencer extends CValueSequencer {
  private static m_nCurrentSequence = 0;
  private static m_nCurrentRatio = 0;
  private static m_nCurrentLabel = 0;

  protected m_nYinLine = 2;
  protected m_nYangLine = 1;

  constructor(nValue: number) {
    super(3, 8, nValue);
    this.m_pvsInner = [
      new CLineValueSequencer(0),
      new CLineValueSequencer(0),
      new CLineValueSequencer(0),
    ];
    this.line(2).setParent(this);
    this.line(1).setParent(this);
    this.line(0).setParent(this);
    this.m_nSequences = Sequences.nTrigramSequences;
    this.m_nRatios = Sequences.nTrigramRatios;
  }

  public line(nIndex: number): CLineValueSequencer {
    return this.m_pvsInner[nIndex] as CLineValueSequencer;
  }

  public override inverse(): this {
    const bMoving = [false, false, false];
    this.saveMoving(bMoving);
    this.value =
      (this.line(2).value % 2 === 0 ? 0 : 1) +
      (this.line(1).value % 2 === 0 ? 0 : 1) * 2 +
      (this.line(0).value % 2 === 0 ? 0 : 1) * 4;
    this.updateInnerValues();
    this.updateOuterValues();
    this.restoreMoving(bMoving, true);
    return this;
  }

  public override opposite(): this {
    const bMoving = [false, false, false];
    this.saveMoving(bMoving);
    super.opposite();
    this.restoreMoving(bMoving, false);
    return this;
  }

  public override move(): this {
    this.line(2).move();
    this.line(1).move();
    this.line(0).move();
    return this;
  }

  public override young(): this {
    this.line(2).young();
    this.line(1).young();
    this.line(0).young();
    return this;
  }

  public override updateInnerValues(): void {
    switch (this.m_nValue) {
      case 0:
        this.line(2).value = this.m_nYinLine;
        this.line(1).value = this.m_nYinLine;
        this.line(0).value = this.m_nYinLine;
        break;
      case 1:
        this.line(2).value = this.m_nYinLine;
        this.line(1).value = this.m_nYinLine;
        this.line(0).value = this.m_nYangLine;
        break;
      case 2:
        this.line(2).value = this.m_nYinLine;
        this.line(1).value = this.m_nYangLine;
        this.line(0).value = this.m_nYinLine;
        break;
      case 3:
        this.line(2).value = this.m_nYinLine;
        this.line(1).value = this.m_nYangLine;
        this.line(0).value = this.m_nYangLine;
        break;
      case 4:
        this.line(2).value = this.m_nYangLine;
        this.line(1).value = this.m_nYinLine;
        this.line(0).value = this.m_nYinLine;
        break;
      case 5:
        this.line(2).value = this.m_nYangLine;
        this.line(1).value = this.m_nYinLine;
        this.line(0).value = this.m_nYangLine;
        break;
      case 6:
        this.line(2).value = this.m_nYangLine;
        this.line(1).value = this.m_nYangLine;
        this.line(0).value = this.m_nYinLine;
        break;
      case 7:
        this.line(2).value = this.m_nYangLine;
        this.line(1).value = this.m_nYangLine;
        this.line(0).value = this.m_nYangLine;
        break;
    }
    this.line(2).updateInnerValues();
    this.line(1).updateInnerValues();
    this.line(0).updateInnerValues();
  }

  public override updateOuterValues(): void {
    this.value =
      (this.line(0).value % 2 === 0 ? 0 : 1) +
      (this.line(1).value % 2 === 0 ? 0 : 1) * 2 +
      (this.line(2).value % 2 === 0 ? 0 : 1) * 4;
    this.m_pvsParent?.updateOuterValues();
  }

  private saveMoving(bMoving: boolean[]): void {
    for (let l = 0; l < 3; ++l) {
      const line = this.line(l);
      if (line.value === 0 || line.value === 3) {
        bMoving[l] = true;
      }
    }
  }

  private restoreMoving(bMoving: boolean[], bInverseLine: boolean): void {
    for (let l = 0; l < 3; ++l) {
      if (bMoving[l]) {
        let l1 = l;
        if (bInverseLine) {
          if (l1 === 0) {
            l1 = 2;
          } else if (l1 === 2) {
            l1 = 0;
          }
        }
        this.line(l1).old();
      }
    }
  }

  public static setCurrentSequence(nSequence: number): void {
    this.m_nCurrentSequence = nSequence;
  }

  public static setCurrentRatio(nRatio: number): void {
    this.m_nCurrentRatio = nRatio;
  }

  public static setCurrentLabel(nLabel: number): void {
    this.m_nCurrentLabel = nLabel;
  }

  protected override getLabel(): string {
    return Sequences.strTrigramLabels[this.getCurrentLabel()][this.value];
  }

  protected override getMoving(): boolean {
    return this.line(0).isMoving || this.line(1).isMoving || this.line(2).isMoving;
  }

  protected override getCurrentSequence(): number {
    return CTrigramValueSequencer.m_nCurrentSequence;
  }

  protected override getCurrentRatio(): number {
    return CTrigramValueSequencer.m_nCurrentRatio;
  }

  protected override getCurrentLabel(): number {
    return CTrigramValueSequencer.m_nCurrentLabel;
  }
}
